Mathlib/Data/Matrix/PEquiv.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2019 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+
+! This file was ported from Lean 3 source module data.matrix.pequiv
+! leanprover-community/mathlib commit 3e068ece210655b7b9a9477c3aff38a492400aa1
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Matrix.Basic
+import Mathlib.Data.PEquiv
+
+/-!
+# partial equivalences for matrices
+
+Using partial equivalences to represent matrices.
+This file introduces the function `PEquiv.toMatrix`, which returns a matrix containing ones and
+zeros. For any partial equivalence `f`, `f.toMatrix i j = 1 ↔ f i = some j`.
+
+The following important properties of this function are proved
+`toMatrix_trans : (f.trans g).toMatrix = f.toMatrix ⬝ g.toMatrix`
+`toMatrix_symm  : f.symm.toMatrix = f.toMatrixᵀ`
+`toMatrix_refl : (PEquiv.refl n).toMatrix = 1`
+`toMatrix_bot : ⊥.toMatrix = 0`
+
+This theory gives the matrix representation of projection linear maps, and their right inverses.
+For example, the matrix `(single (0 : Fin 1) (i : Fin n)).toMatrix` corresponds to the ith
+projection map from R^n to R.
+
+Any injective function `Fin m → Fin n` gives rise to a `PEquiv`, whose matrix is the projection
+map from R^m → R^n represented by the same function. The transpose of this matrix is the right
+inverse of this map, sending anything not in the image to zero.
+
+## notations
+
+This file uses the notation ` ⬝ ` for `Matrix.mul` and `ᵀ` for `Matrix.transpose`.
+-/
+
+
+namespace PEquiv
+
+open Matrix
+
+universe u v
+
+variable {k l m n : Type _}
+
+variable {α : Type v}
+
+open Matrix
+
+-- Porting note: added
+local instance [DecidableEq n] (j : n) (o : Option n) : Decidable (j ∈ o) :=
+  haveI : Decidable (o = some j) := inferInstance
+  this
```

Comment:
This is definitely better, but belongs in the same file that defines the `∈` notation

Reply:
That's Std.Data.Option.Basic. I'll submit a PR there tomorrow, but wouldn't mind if we moved this along in the meantime.